### ISSUE TITLE
[Discussion] Add the ability to handle images in Onyx

### ISSUE THREAD

**Issue by mallenexpensify** [Expensify/App#3867]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
# Problem
Right now, we have no way of saving images offline.  We can't use Onyx as it has storage limitations on web with a 5MB limit.  There's been a lot of discussion on how to handle this in this [issue](https://github.com/Expensify/Expensify.cash/issues/2258) and in this [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1625092068290500).  

# Why is this important?
This is important to be able to handle offline functionality of images

# Solution
Create an offline storage solution for Onyx so that even when the application is offline and closed, the attached images will persist and continue to upload

**Expensify/Expensify Issue URL:**
https://github.com/Expensify/Expensify/issues/169197
cc @thienlnam 

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @greg-schroeder (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by parasharrajat** [comment]
> Create a JNI-based AsyncStorage for Onyx that works across all platforms

Sorry but this looks like a huge task to do. Maybe we did not mean to do so. Can anybody confirm what exactly is need to be built? I expect that we need to expand AsyncStorage in a wrapper so that it supports a method like `AsyncStorage.setImages()` or `AsyncStorage.removeImages()`. Please, this looks daunting to me.

**Comment by thienlnam** [comment]
@parasharrajat Yup this is a huge task, but basically is going to be summed up with what needs to happen from this[ slack tread](https://expensify.slack.com/archives/C01GTK53T8Q/p1625092068290500) happening now.  I updated the solution a bit but I imagine that it will keep updating until we figure out what the most viable solution for this will be

**Comment by mallenexpensify** [comment]
Posted to https://www.upwork.com/jobs/~011fcd226e59956a18

I'm going to add a new Contributor Manager, I'm out of office next week. They can update the price if needed (sounds like @parasharrajat might want for $250 though :trollface: )

**Comment by MelvinBot** [comment]
Triggered auto assignment to @SofiedeVreese (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by parasharrajat** [comment]
@mallenexpensify :jack_o_lantern:. Maybe I will find a solution that takes lesser effort.

**Comment by aditya27dev** [comment]
@mallenexpensify I have applied on upwork by the name "Aditya Batra" on your JD.

**Comment by SofiedeVreese** [comment]
Hey @aditya27dev can you a proposal with the technical explanation of the changes you will make in this GH here? thanks!

**Comment by aditya27dev** [comment]
Hi @SofiedeVreese Sorry I can't share technical changes here publicly which is a inappropriate way of doing a job.

But I can work on this task, if you hire me.

**Comment by SofiedeVreese** [comment]
Hey @aditya27dev per our [contributing guidelines](https://github.com/Expensify/Expensify.cash/blob/main/CONTRIBUTING.md), we do require contributors to make a proposal for their solution and post it as a comment in the corresponding GitHub issue. The solution proposal should include a brief technical explanation of the changes proposed. We don't hire if the contributing guidelines aren't being followed.

**Comment by thienlnam** [comment]
Throwing a new title on this since discussion is still ongoing in the slack thread, to view or participate check it out [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1625092068290500)

**Comment by SofiedeVreese** [comment]
Seems like the discussion is still fairly ongoing.

**Comment by SofiedeVreese** [comment]
Clarifying with @thienlnam to see if there's some sort of resolution in the Slack thread.

**Comment by SofiedeVreese** [comment]
Adding a planning label and changing to a weekly for now, per [this comment](https://github.com/Expensify/Expensify/issues/169197#issuecomment-878586906). We're still discussing possible solutions so having this as a daily doesn't make sense for now.

**Comment by SofiedeVreese** [comment]
I don't think we have any updates here. @thienlnam should we close this GH or is there still an ongoing discussion? I was assigned as part of the "external" label (= to push it to Upwork) but I'm not sure if my assignment is still relevant here.

**Comment by thienlnam** [comment]
Discussion has mainly halted - here is where we're at with the last comment from @kidroca 

> Neither Android or iOS are handling file/blob pasting in E.cash ATM
> On web/desktop the clipboard provides a File object
> The only time something happens from a blob is when html content is copied, then any image sources are fetched as blobs, and then an in memory File is constructed. The preview image uses the File.sourceURL
> The File constructor is available only on web/desktop not on native
> It might be possible to do the same on native, but would involve more hoops than simply using the FS
> And I can't even find anything on the subject, besides converting the blob to a base64 string and feeding it as image source.uri
> So you have to:
> Read blob from SQL and keep it in memory (as less as possible)
> convert it to something the Image can work with.
> Feed it to the Image component
> Compared to:
> Have a URI string pointing to a file in FS
> Give it to the Image
> ---
> Before making this decision let's find an example that uses Blobs to display images and what it involves. E.g. if it have to store the blob on disc first, even temporary, it defeats the purpose

Going to be looking into this in the next couple of weeks but if someone else would like to dig into this feel free!

**Comment by thienlnam** [comment]
Been focused on a few other things at the moment so I haven't had time to look into this yet - the investigation can still be an external task

**Comment by arielgreen** [comment]
Hey, @kidroca, can you spend some time investigating this? You can do this on your existing contract

**Comment by kidroca** [comment]
@arielgreen Yes, I've taken part of the discussion on slack and can research this further

**Comment by kidroca** [comment]
The original plan was to keep image/files in the database. This is easy to achieve on web/desktop, but more complicated on native with the current storage library and model we use
1. AsyncStorage works only with strings ATM - converting files to base64 strings should be out of the question. But we can discuss it if you want to
2. MMKV is no better as it only supports simple types like number, boolean, or string

Trying to store images/files in the database would be tremendously difficult even if we switch to a richer database library 
I've tried to explain it on slack. 
The key point for storing everything in "local db" is simplicity and universal logic, but that won't be the case.

I'll propose a simpler and IMO more maintainable solution.

---

Onyx addresses the storage layer through a simple get/set/merge interface 
We've already extracted this interface and were able to swap different underlying storage "drivers" that can be selected based on platform (e.g. MMKV for native and AsyncStorage for web/desktop)

### iOS and Android

Let's say we have a `storage/index.native.js` 
Onyx imports `Storage` from `./storage` and uses it like `Storage.get`, `Storage.set`, etc...
Under the hood it delegates calls to `AsyncStorage` like we already do
While for the file related operations we'd use `react-native-fs`
We can either add `getFile/setFile` specific methods or overload the existing `get`, `set` methods to work with files based on input type

#### Storing files
A request to save a file would be very similar to existing storage calls - it will have a `key` and `value` 
The `key` should be based on the `remote` url of the file (if such exists) and can be prefixed by a collection name 
e.g. `FILES_${key}` 
The `value` should be `File` metadata (this is provided by our file/image picker) 
Using `react-native-fs` we store the file in our app documents (or cache) space with a file name that's based on the `key` e.g. `FILES_${key}` 
The specific implementation can either move or copy the file there
As with regular storage, we should have eviction logic that cleans up files when storage is low or the OS complains

Storing files this way is very fast - we're working with a reference to the file and it never actually travels from JS to native
Moving/copying operations are handled by the OS and threaded separately

#### Reading files
Reading is based on the same concept
the Onyx API receives a key like `FILES_${key}` it's identified as a request to read a file 
We know the location of the file because it's based on the key and the storage folder
Using `react-native-fs` we create a `File` reference to the file
Images and File preview components can use this to render the file 

Accessing the file in this manner means
1. We're **not** reading the whole file from native to JS and then keep it in memory
2. The preview is loaded only when it has to and it's handled natively

You can checkout how the "storage provider" concept works in practice here: https://github.com/Expensify/react-native-onyx/pull/95

### Web and Desktop

Let's say we have a `storage/index.js` 
Onyx imports `Storage` from `./storage` and uses it like `Storage.get`, `Storage.set`, etc...
Under the hood it delegates calls to `AsyncStorage` like we already do
But for the file related operations we'll implement a solution that stores Files in IndexedDb as `Blob`s 

#### Storing files
The flow is very similar to the one for native. 
We'll be using a key like `FILES_${key}` and `File` metadata
Then we'll convert the file to `Blob` and save it under the key in `IndexedDB` 

#### Reading files
We receive a key which we evaluate as a File key 
We use it to get a `Blob` from the database
When we retrieve the `Blob` we can convert it to `File` using the `File` constructor - our web/desktop implementations have similar logic for pasted images (see `TextInputFocusable/index.js`) 
Images and File preview components can use the file to render

---

Using different Storage Provider for native and web allows us to implement the best storage handling for each platform. 
Even though we're using Platform specific code, it's encapsulated and abstracted by Onyx so that Onyx user like Expensify/App don't have to

**Comment by thienlnam** [comment]
@kidroca This proposal looks great - I agree that we should be using the best storage handling for each platform and encapsulating it through Oynx.  If possible, let's explore this with a POC and see how this would work within Oynx

**Comment by kidroca** [comment]
Alright, I'll prepare a POC that stores and displays an image with Onyx

**Comment by kidroca** [comment]
Submitted 2 PRs 

1. Add file handling capabilities to Onyx: https://github.com/Expensify/react-native-onyx/pull/102
2. Small updates to Expensify/App to showcase offline file persistence: https://github.com/Expensify/App/pull/4908

You can also use the branch to build and run the POC: https://github.com/kidroca/Expensify.cash/tree/kidroca/file-handling-poc

#### Overview of the changes

The changes are based on the idea here (though not exactly): https://github.com/Expensify/App/issues/3867#issuecomment-902662445

It was easy to move existing web/desktop handling to happen through [localforage](https://localforage.github.io/localForage/), this allows us to save `FIles` in IndexedDB with almost no changes. Using `localforage` moves all key/value handling to IndexedDB which allows much more than 5MB for everything that we store in Onyx

On native we store a reference to a file, this is OK as everything working with native files supports this and expects this format. The only thing that we do is to move/copy the file to a place where we know it would not be removed by the user or the OS. This is the `app/documents/onxy_files` folder

There's still some work to be done if the POC is accepted (eviction, error cases), the POC tries to showcase this functionality:
> Create an offline storage solution for Onyx so that even when the application is offline and closed, the attached images will persist and continue to upload

### What needs to be done 
#### File cleanup logic
Clean up works on web/desktop as removing the persisted request removes the `File` with it
But on mobile, the file is kept separately and we need to relay to Onyx that a file is no longer needed

Perhaps we can update the Network queue to call a function like `Onyx.releaseFile(file)` when a request containing a file
was completed.
A simpler solution might be to run a check during launch (or before exit) that removes files older than one week, assuming they were uploaded by our request retry logic

#### Eviction logic
Handle errors resulting from lack of space
There's functionality that lists the files created by "us". We can add filtering and remove the oldest ones

#### Cleanup and documentation
There are some "Todo" comments on possible enhancements or alterations
Some items would need to be better documented

**Comment by thienlnam** [comment]
Awesome, this is looking really good - going to loop in @marcaaron and @tgolen and see what their thoughts on the POC are

**Comment by thienlnam** [comment]
No update here yet - going to start pinging people for some thoughts / reviews on the POC

**Comment by kidroca** [comment]
When you open the PR you see this 
![image](https://user-images.githubusercontent.com/12156624/132894429-e455e593-7c65-41d1-b65a-c940e476e025.png)
which might be scary 👻 but it's just the package-lock update
![image](https://user-images.githubusercontent.com/12156624/132894527-0423e0aa-3b05-4204-b076-3370a9c2d91e.png)

**Comment by thienlnam** [comment]
Pinged for reviews but n6 deadline is taking priority - going to ping about this again next week

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by kidroca** [comment]
We've made some changes in the Onyx PR: https://github.com/Expensify/react-native-onyx/pull/102
The file handling process for native is simplified a lot and it makes usage easier. There are notes in the description: 
> When an attachment is picked the `AttahcmentPicker` will persist it to App/cache folder. It is then available until cache is cleared
This lets us save only the reference to the File in our key/val storage
This reference is enough to work with the file - use it in an `<Image />` or in a upload request

The PR is approved, but we might need to deal with an issue on mobile web ios first.

**Comment by thienlnam** [comment]
Discussions have been happening in the relevant PRs

**Comment by quinthar** [comment]
Out of curiosity, when we add E2E encryption, how do we intend for this to work?  I think we can maintain it unencrypted on disk such that it works in the `<img>` tag, but when we upload it will it be part of the regular message body as normal?

**Comment by kidroca** [comment]
There's an issue on mobile-web due to us deleting persisted requests (and their files) as soon as they are added to the network queue: https://github.com/Expensify/App/pull/4908#issuecomment-967139238

The work in this ticket: https://github.com/Expensify/App/issues/5987 would also solve the problem here as it will only remove persisted data after the request is completed

**Comment by thienlnam** [comment]
Looks like PR is ready for review https://github.com/Expensify/App/issues/5987
Once that is merged, we can merge the other file handling PR https://github.com/Expensify/react-native-onyx/pull/102

**Comment by kidroca** [comment]
Synced with `main` (1170 commits 🤯)
I've reverted the change to persist attachment requests to storage so that we can merge this as it was [approved](https://github.com/Expensify/react-native-onyx/pull/102#pullrequestreview-797147352), and once #6556 is merged we can just enable persisting attachment requests

**Comment by roryabraham** [comment]
I understand the Onyx POC was just merged and that we want a follow-up to add some more tests. Is there an issue somewhere to bring these changes to life and implement offline-first images in the New Expensify app?

**Comment by thienlnam** [comment]
Yeah so this 
PR: https://github.com/Expensify/App/pull/4908#event-5768021643 (I've been meaning to review) is so that we can handle images with onyx 

After that, this issue (https://github.com/Expensify/App/issues/6527) is probably the next step for offline-first images

**Comment by kidroca** [comment]
One thing I forgot to bring up 
When #4908 it will switch `web/desktop` from using one storage medium - localStorage to another - IndexedDB
Meaning any logged in users would be logged out

Should we include a migration that tries to move any existing Onyx values in localStorage to IndexedDB before the app starts (along the rest of the migrations)? 

ios and android aren't going to be affected since they continue to use the same medium - AsyncStorage

**Comment by parasharrajat** [comment]
I guess yes, we should add that.

**Comment by tgolen** [comment]
Yeah, I agree, let's try to migrate everything if we can.

**Comment by kidroca** [comment]
Alright, I'm working on it

**Comment by kidroca** [comment]
Added the migration: https://github.com/Expensify/App/pull/4908/files#diff-fc59a9d775dba3004d8ed5a4dec154bbd2487455e7c720afed29bc8003a5228f

**Comment by kidroca** [comment]
PR was approved but we forgot to remove the HOLD label: https://github.com/Expensify/App/pull/4908#issuecomment-1010993484

**Comment by kidroca** [comment]
We can enable file persistence after #6556 is merged 

It deals with the issue of prematurely deleting persisted request (and the files to be uploaded) 

After #6556 we'll be able to remove the comment here and set `persist` to `true` 

https://github.com/Expensify/App/blob/c224cfe5ba9f9fc2214a2963d6c02d855dc1c96e/src/libs/actions/Report.js#L1170-L1173

**Comment by kidroca** [comment]
Still waiting on #6556

**Comment by kidroca** [comment]
#6556 was merged 

Created a PR to enable attachment persistence: #7520

**Comment by kidroca** [comment]
Found 2 problems around persisted attachments

### No loading animation/preview
You can see it on most of the videos on the PR #7520 (see web or mobile web) 

1. Be offline 
2. Select an attachment to be uploaded (preferable larger so it takes more time to upload)
3. Quit the app
4. Open the app again
5. Open the chat where the attachment was uploaded
6. Observe it's not there, and there's no indication something is uploading
7. The attachment only appears after the upload is over 

#### Problem:
The optimistic action that serves as a placeholder during the upload is removed 
We seem to clear any optimistic actions at launch or when a chat is opened
The reason seem to be to prevent discrepancies between local and online state - any optimistic actions are removed - if real actions exist for the same they would be fetched shortly 
https://github.com/Expensify/App/blob/425e718b9ed3ac485da1e86a11779941daa95ba5/src/libs/actions/Report.js#L451-L457

<details>
<summary>Solution</summary>
I think we should try to cross reference optimistic actions with requests in the network queue
This way when we clean up optimistic actions we keep the ones that still have pending request

We already put `optimisticReportActionID` on the request so we can distinct it
https://github.com/Expensify/App/blob/425e718b9ed3ac485da1e86a11779941daa95ba5/src/libs/actions/Report.js#L1169-L1175

We can make a lib or expose a function that return any pending `Report_AddComment` actions from network queue or persisted storage
`Report.removeOptimisticActions` would use that and skip removing optimistic actions that have pending `Report_AddComment` 
</details>


### 2. We're only persisting a request if we're offline 

`NetworkRequestQueue.saveRetryableRequests` is only called in the `if(isOffline)` block 

https://github.com/Expensify/App/blob/425e718b9ed3ac485da1e86a11779941daa95ba5/src/libs/Network.js#L185-L206

#### Problem: 
If we quit the app while an upload is still pending it will neither be upload or persisted
For a request to be persisted you have to be offline first and then to make the request

[Recent changes](https://github.com/Expensify/App/pull/6556) alleviate this to some extent - when a request is persisted to storage it will only be removed after it's successfully performed - it'll be retried until eventually succeeding
But it would only get to be persisted if you were offline to begin with

<details>
<summary>Solution</summary>
Requests with the `persist` flag should be persisted regardless of being offline or not
If the user quits the app while the request isn't completed existing logic would retry it on next launch
Otherwise added logic would remove it from persistent storage when it succeeds 

The following
https://github.com/Expensify/App/blob/425e718b9ed3ac485da1e86a11779941daa95ba5/src/libs/Network.js#L229-L230

becomes
```js
 if (queuedRequest.data.persist) {
   NetworkRequestQueue.saveRetryableRequests([queuedRequest]);
 }

 processRequest(queuedRequest) 
     .then(response => {
        onResponse(queuedRequest, response);
        NetworkRequestQueue.removeRetryableRequest(queuedRequest);
     }) 
```
</details>

**Comment by thienlnam** [comment]
We should be live with this already and can close this right? @kidroca

**Comment by kidroca** [comment]
@thienlnam I think we have the bare minimum if that's enough I guess we can close 

I've pointed out [2 problems](https://github.com/Expensify/App/issues/3867#issuecomment-1028183382) using the feature the way it is right now

**Comment by thienlnam** [comment]
👍  Will keep this open until we resolve those issue

**Comment by thienlnam** [comment]
I'll create separate tickets for them for more visibility and then we can discuss the proposed solutions in those tickets

**Comment by thienlnam** [comment]
Closing now that we have other tickets for the existing problems
https://github.com/Expensify/App/issues/7934
https://github.com/Expensify/App/issues/7935
